MangoAccount create from AccountInfo

bdhobare opened the pull request:
- Added `MangoAccountInfo` for the decoded information from RPC API and `MangoAccount` that holds more methods like account health, liquidation status, etc  that can be easily implemented and tested in subsequent PRs.
- Added two ways to create the `MangoAccount`:  using prefetched `AccountInfo` or by fetching it directly from RPC during construction.
- Tests for these 2 methods and class
```cpp
// option 1. Fetch info implicitly from RPC
const auto& mangoAccount = mango_v3::MangoAccount::from(solana::PublicKey::fromBase58(key), mango_v3::DEVNET.endpoint);
auto liquidationPrice = mangoAccount->getLiquidationPrice().toDouble();

// option 2 : Use prefetched AccountInfo
const auto& mangoAccountInfo =
      connection.getAccountInfo<mango_v3::MangoAccountInfo>(pubKey);
const auto& mangoAccount = mango_v3::MangoAccount::from(mangoAccountInfo);

```

mschneider commented on tests/main.cpp:
```diff
@@ -97,3 +97,21 @@ TEST_CASE("Test getLatestBlock") {
+  const auto& mangoAccountInfo_ =
+      connection.getAccountInfo<mango_v3::MangoAccountInfo>(key);
+  const auto& mangoAccount = mango_v3::MangoAccount::from(mangoAccountInfo_);
+  CHECK(!std::is_null_pointer<decltype(mangoAccount)>::value);
```
i'm not sure what this check does

bdhobare replied:
The returned `unique_ptr` isn't a `nullptr_t`. Unlikely,  that it will be null without an exception happening( e.g some copy failed) which will just terminate the test but wanted to just be extra strict and validate everything.

mschneider replied:
yeah but this is a type check, the type checker doesn't validate runtime errors

mschneider replied:
you need to use CHECK(mangoAccount) to verify it's not null

mschneider commented on include/mango_v3.hpp:
```diff
@@ -125,7 +127,7 @@ struct MangoAccount {
   i80f48 borrows[MAX_TOKENS];
   solana::PublicKey spotOpenOrders[MAX_PAIRS];
-  PerpAccount perpAccounts[MAX_PAIRS];
+  PerpAccountInfo perpAccounts[MAX_PAIRS];
```
yep, good call, you found an inconsistency here in the upstream code. we should make this consistent in mango v4

mschneider commented on include/mango_v3.hpp:
```diff
@@ -140,6 +142,39 @@ struct MangoAccount {
+  i80f48 getLiquidationPrice() { return {}; }
+  // TODO: Add methods to calculate health ratio
+ private:
+  const MangoAccountInfo &accountInfo_;
```
don't think it's good that this one is const and private. i know that a lot of people claim that you should always keep everything and private and so on, but i really hate it, when it limits the usability of the code. 

it's really handy to be able to create a copy and manipulate some values in there to simulate operations like withdrawal margin checks client side. we do it frequently in mango ui.

bdhobare replied:
Will update this to remove this `const` for now. I will add a public method to retrieve  the `AccountInfo` of a given `MangoAccount`

mschneider replied:
try to use as little encapsulation as necessary and not default to the maximum. it just creates code that no one wants to read.

mschneider commented on include/mango_v3.hpp:
```diff
@@ -140,6 +142,39 @@ struct MangoAccount {
+    return std::make_unique<MangoAccount>(NotPubliclyConstructible{},
+                                          accountInfo);
+  }
+  i80f48 getLiquidationPrice() { return {}; }
```
there is no method like that in mango unfortunately

bdhobare replied:
Great explanation, I didn't know this. I was trying to replicate the methods provided by the Typescript SDK in `MangoAccount.ts` but they don't explain these internal details. This method doesn't do anything for now but this explanation is great for future implementation.

mschneider commented on include/mango_v3.hpp:
```diff
@@ -140,6 +142,39 @@ struct MangoAccount {
+    auto connection = solana::rpc::Connection(endpoint);
+    const auto &accountInfo =
+        connection.getAccountInfo<MangoAccountInfo>(pubKey.toBase58());
+    return std::make_unique<MangoAccount>(NotPubliclyConstructible{},
```
unique_ptr with reference to rvalue? uuh i'm not sure this won't segfault

mschneider replied:
also I try to call toBase58() in any internal method, strictly only in examples as it does come at considerable cost which is bad for latency, rather try to design the interfaces that the user needs to call it and hence can decide to cache the result.

bdhobare replied:
> unique_ptr with reference to rvalue?

Totally missed this, thanks for noticing it.

mschneider replied:
Unless you need to use the heap, try not to introduce unique ptrs. Best left to the users to decide, when they want to work on stack and when on heap.

bdhobare replied:
> as it does come at considerable cost which is bad for latency

I will change the `PubKey` parameter to a `string` ,  just like `getAccountInfo` as that is what it is used for in this method really.

mschneider commented on include/mango_v3.hpp:
```diff
@@ -140,6 +142,39 @@ struct MangoAccount {
+  // Creating a MangoAccount might involve non-trivial work like calling RPC
+  // methods, so we provide explict `MangoAccount::from()` overloaded set of
+  // functions and make the public constructor un-callable.
+  struct NotPubliclyConstructible {};
```
dont buy the argument, same could be achieved by just setting the default constructor private, as even rpc methods only use exceptions for error handling, they are fine to be called from constructors.

bdhobare commented:
@mschneider Simplified `MangoAccount` to a struct that will have the other methods added. Also replaced the explicit initialization methods with normal constructors. Updated the tests to validate the accountInfo instead of the type.
